src/proper_erlang_abstract_code.erl

```diff
@@ -0,0 +1,3175 @@
+pfun(pat_var, _) -> fun pat_var/1;
+pfun(plain_list, Where) -> pfun1(fun plain_list/2, Where);
+pfun('receive', _) -> fun 'receive'/1;
+pfun(rec_lambda, _) -> fun rec_lambda/1;
+pfun(record, Where) -> fun(S) -> a_record(S, Where) end;
+pfun(build_record, Where) -> fun(S) -> build_record(S, Where) end;
+pfun(record_pattern, _) -> fun record_pattern/1;
+pfun(update_record, Where) -> pfun1(fun update_record/2, Where);
+pfun(record_index, _) -> fun(S) -> record_index(S) end;
+pfun(record_field_access, Where) -> pfun1(fun record_field_access/2, Where);
+pfun(extcall, _) -> fun extcall/1;
+pfun(remote_guard_call, _) -> fun remote_guard_call/1;
+pfun(small, Where) -> fun(S) -> small(S, Where) end;
+pfun(string, _) -> fun a_string/1;
+pfun(termcall, Where) -> pfun1(fun termcall/2, Where);
+pfun(tuple, Where) -> fun(S) -> tuple(S, Where) end;
+pfun('try', _) -> fun 'try'/1;
+pfun(try_of, _) -> fun try_of/1;
+pfun(try_after, _) -> fun try_after/1;
+pfun(unop, Where) -> fun(S) -> unop(S, Where) end;
+pfun(update_map, Where) -> pfun1(fun update_map/2, Where);
+pfun(var, _) -> fun var/1;
+pfun(varcall, Where) -> pfun1(fun varcall/2, Where);
+pfun(var_eclass, _) -> fun var_eclass/1;
+pfun(var_timeout, _) -> fun var_timeout/1;
+pfun(yes_guard, _) -> fun yes_guard/1;
+pfun(yes_multi_field_init, _) -> fun yes_multi_field_init/1.
+
+pfun1(F, Where) -> fun(S) -> F(S, where(S, Where)) end.
+
+where(S, compound) -> abstract_expr(S);
+where(S, guard_test) -> guard_test(S);
+where(S, pattern) -> pattern(S).
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+options(Options) ->
+    State = create_template(),
+    #gen_state{weights = Weights, limits = Limits} = State,
+    (check_options(Weights, Limits, Options)
+     andalso eval_options(Options, State)).
+
+check_options(Weights, Limits, Options) ->
+    lists:all(fun ({K, {V1, V2}}) -> check_option2(Weights, Limits, K, V1, V2)
+                ; ({K, V}) -> check_option(K, V)
+              end, Options).
+
+check_option(set_all_weights, Value) ->
+    is_integer(Value) andalso Value >= 0;
+check_option(variables, Term) ->
+    try
+        %% true = length(Term) > 0,
+        lists:all(fun(V) -> is_variable(V) end, Term)
+    catch
+        _:_ ->
+            false
+    end;
+check_option(functions, Term) ->
+    try
+        lists:all(fun(F) -> is_fa(F) end, Term)
+    catch
+        _:_ ->
+            false
+    end;
+check_option(types, Term) ->
+    try
+        lists:all(fun(T) -> is_fa(T) end, Term)
+    catch
+        _:_ ->
+            false
+    end;
+check_option(records, Term) ->
+    try
+        lists:all(fun(R) -> is_record(R) end, Term)
+    catch
+        _:_ ->
+            false
+    end;
+check_option(char, Term) when is_function(Term, 0) ->
+    true;
+check_option(atom, Term) when is_function(Term, 0) ->
+    true;
+check_option(resize, Term) when Term; not Term ->
+    true;
+check_option(_, _) ->
+    false.
+
+check_option2(Weights, _Limits, weight, Term1, W) ->
+    try
```

Comment:
Wouldn't this be simpler without the `try-catch` but with an added `is_map(Weights)` call?